// react-vega/lib/Vega.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _shallowEqual = _interopRequireDefault(require("./utils/shallowEqual"));

var _updateMultipleDatasetsInView = _interopRequireDefault(require("./utils/updateMultipleDatasetsInView"));

var _VegaEmbed = _interopRequireDefault(require("./VegaEmbed"));

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const EMPTY = {};

class Vega extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "vegaEmbed", /*#__PURE__*/_react.default.createRef());

    _defineProperty(this, "handleNewView", view => {
      this.update();
      const {
        onNewView = _constants.NOOP
      } = this.props;
      onNewView(view);
    });
  }

  componentDidMount() {
    this.update();
  }

  componentDidUpdate(prevProps) {
    if (!(0, _shallowEqual.default)(this.props.data, prevProps.data)) {
      this.update();
    }
  }

  update() {
    const {
      data
    } = this.props;

    if (this.vegaEmbed.current && data && Object.keys(data).length > 0) {
      this.vegaEmbed.current.modifyView(view => {
        (0, _updateMultipleDatasetsInView.default)(view, data);
        view.resize().run();
      });
    }
  }

  render() {
    const {
      data,
      ...restProps
    } = this.props;
    return /*#__PURE__*/_react.default.createElement(_VegaEmbed.default, _extends({
      ref: this.vegaEmbed
    }, restProps, {
      onNewView: this.handleNewView
    }));
  }

}

exports.default = Vega;

_defineProperty(Vega, "defaultProps", {
  data: EMPTY
});
