// react-vega/lib/VegaEmbed.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _vegaEmbed = _interopRequireDefault(require("vega-embed"));

var _shallowEqual = _interopRequireDefault(require("./utils/shallowEqual"));

var _getUniqueFieldNames = _interopRequireDefault(require("./utils/getUniqueFieldNames"));

var _constants = require("./constants");

var _addSignalListenersToView = _interopRequireDefault(require("./utils/addSignalListenersToView"));

var _computeSpecChanges = _interopRequireDefault(require("./utils/computeSpecChanges"));

var _removeSignalListenersFromView = _interopRequireDefault(require("./utils/removeSignalListenersFromView"));

var _combineSpecWithDimension = _interopRequireDefault(require("./utils/combineSpecWithDimension"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class VegaEmbed extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "containerRef", /*#__PURE__*/_react.default.createRef());

    _defineProperty(this, "resultPromise", void 0);

    _defineProperty(this, "handleError", error => {
      const {
        onError = _constants.NOOP
      } = this.props;
      onError(error, this.containerRef.current); // eslint-disable-next-line no-console

      console.warn(error);
      return undefined;
    });

    _defineProperty(this, "modifyView", action => {
      if (this.resultPromise) {
        this.resultPromise.then(result => {
          if (result) {
            action(result.view);
          }

          return true;
        }).catch(this.handleError);
      }
    });
  }

  componentDidMount() {
    this.createView();
  }

  componentDidUpdate(prevProps) {
    const fieldSet = (0, _getUniqueFieldNames.default)([this.props, prevProps]);
    fieldSet.delete('className');
    fieldSet.delete('signalListeners');
    fieldSet.delete('spec');
    fieldSet.delete('style');
    fieldSet.delete('width');
    fieldSet.delete('height'); // Only create a new view if necessary

    if (Array.from(fieldSet).some(f => this.props[f] !== prevProps[f])) {
      this.clearView();
      this.createView();
    } else {
      const specChanges = (0, _computeSpecChanges.default)((0, _combineSpecWithDimension.default)(this.props), (0, _combineSpecWithDimension.default)(prevProps));
      const {
        signalListeners: newSignalListeners
      } = this.props;
      const {
        signalListeners: oldSignalListeners
      } = prevProps;

      if (specChanges) {
        if (specChanges.isExpensive) {
          this.clearView();
          this.createView();
        } else {
          const areSignalListenersChanged = !(0, _shallowEqual.default)(newSignalListeners, oldSignalListeners);
          this.modifyView(view => {
            if (specChanges.width !== false) {
              view.width(specChanges.width);
            }

            if (specChanges.height !== false) {
              view.height(specChanges.height);
            }

            if (areSignalListenersChanged) {
              if (oldSignalListeners) {
                (0, _removeSignalListenersFromView.default)(view, oldSignalListeners);
              }

              if (newSignalListeners) {
                (0, _addSignalListenersToView.default)(view, newSignalListeners);
              }
            }

            view.run();
          });
        }
      } else if (!(0, _shallowEqual.default)(newSignalListeners, oldSignalListeners)) {
        this.modifyView(view => {
          if (oldSignalListeners) {
            (0, _removeSignalListenersFromView.default)(view, oldSignalListeners);
          }

          if (newSignalListeners) {
            (0, _addSignalListenersToView.default)(view, newSignalListeners);
          }

          view.run();
        });
      }
    }
  }

  componentWillUnmount() {
    this.clearView();
  }

  createView() {
    const {
      spec,
      onNewView,
      signalListeners = {},
      width,
      height,
      ...options
    } = this.props;

    if (this.containerRef.current) {
      const finalSpec = (0, _combineSpecWithDimension.default)(this.props);
      this.resultPromise = (0, _vegaEmbed.default)(this.containerRef.current, finalSpec, options).then(result => {
        if (result) {
          const {
            view
          } = result;

          if ((0, _addSignalListenersToView.default)(view, signalListeners)) {
            view.run();
          }
        }

        return result;
      }).catch(this.handleError);

      if (onNewView) {
        this.modifyView(onNewView);
      }
    }
  }

  clearView() {
    if (this.resultPromise) {
      this.resultPromise.then(result => {
        if (result) {
          result.finalize();
        }
      }).catch(this.handleError);
    }

    this.resultPromise = undefined;
    return this;
  }

  render() {
    const {
      className,
      style
    } = this.props; // Create the container Vega draws inside

    return /*#__PURE__*/_react.default.createElement("div", {
      ref: this.containerRef,
      className: className,
      style: style
    });
  }

}

exports.default = VegaEmbed;

_defineProperty(VegaEmbed, "propTypes", {
  className: _propTypes.default.string,
  onError: _propTypes.default.func
});
